src/reducers/index.js

```diff
@@ -244,16 +244,32 @@ function initializeRowHeightsAndOffsets(state) {
  * @private
  */
 function setStateFromProps(state, props) {
-  const { columnGroupProps, columnProps, elementTemplates, useGroupHeader } =
-    convertColumnElementsToData(props.children);
+  const {
+    columnGroupElements,
+    columnElements,
+    elementTemplates,
+    useGroupHeader,
+  } = convertColumnElementsToData(props.children);
 
   Object.assign(state, {
-    columnGroupProps,
-    columnProps,
+    columnGroupElements,
+    columnElements,
     elementTemplates,
     propsRevision: state.propsRevision + 1,
   });
 
+  // NOTE (pradeep): We pre-freeze these large collections to avoid
+  // performance bottle necks
+  //
+  // From Immer's docs:
+  //     Immer freezes everything recursively. For large data objects
+  //     that won't be changed in the future this might be over-kill,
+  //     in that case it can be more efficient to shallowly
+  //     pre-freeze data using the freeze utility.
+  Object.freeze(state.columnElements);
+  Object.freeze(state.columnGroupElements);
+  Object.freeze(state.elementTemplates);
```

Comment:
Can  `convertColumnElementsToData` itself return these objects as frozen???

Reply:
We could, but I like to keep `convertColumnElementsToData` as a pure helper method that shouldn't know about immer/parent reducer details.